### ISSUE TITLE
Model can't be initialized when np>=2

### ISSUE THREAD

**Issue by joanlynnlove** [uber/horovod#707]
> centos (7.2)
> anaconda (3-5.2.0-Linux-x86_64)
> python (3.6)
> horovod (0.15.2)
> tensorflow-gpu (1.11)
> cuda (9.0)
> cudnn (7.3.1)
> openmpi (4.0.0)
> rsh (Not clear)
> openssh (None)
> nccl (None)

Hi, there. I am quite new to Horovod and MPI. Recently I've been trying to run TensorFlow distributed training in a single **Docker container** with 8 GPUs. **The container has no sshd process running.** 

It works fine on /device:GPU0 when I have only 1 sub process running:

> num_workers=1; mpirun --allow-run-as-root -np $num_workers -H localhost:$num_workers -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib python my_dist_training.py

But when I have 2 (or more) sub processes,  it seems that they can't find each other.

> num_workers=2; mpirun --allow-run-as-root -np $num_workers -H localhost:$num_workers -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib python my_dist_training.py

Some logs are pasted below:

> Adding visible gpu devices: 0
> Adding visible gpu devices: 1
> WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that differen
> t ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock.
> Stalled ops:
> HorovodBroadcast_residual_layer_block_layer1_block_layer1_part_0_block_layer1_part_0_bn_0_batch_normalization_moving_variance_0 [missing ranks: 1]
> HorovodBroadcast_residual_layer_block_layer1_block_layer1_part_0_block_layer1_part_0_bn_0_batch_normalization_beta_0 [missing ranks: 1]
> HorovodBroadcast_residual_layer_block_layer1_block_layer1_part_0_block_layer1_part_0_bn_0_batch_normalization_gamma_0 [missing ranks: 1]
>
> ...

I think this means that rank 0 can't communicate with rank 1. Therefore rank 1 also has no way to init it model parameters:

> INFO:tensorflow:Waiting for model to be ready.  Ready_for_local_init_op:  Variables not initialized: global_step, init_conv_layer/init_conv_0/w, ...

However, I have passed test cases about GPU in horovod-master/test/test_tensorflow.py after I put 'with device('/device:gpu$N')' before 'with tf.session()' in the code snippet.

My core implementation is like:

> gpu_options = tf.GPUOptions(allow_growth=True)
> session_conf = tf.ConfigProto(
> ​        gpu_options=gpu_options,
> ​        allow_soft_placement=True,
> ​        log_device_placement=False,
> )
> global_step = tf.train.get_or_create_global_step()
> self.model = self._get_target_model()
> train_op, grads_and_vars = self._get_grads(self.FLAGS.learning_rate, global_step)
> train_summary_op = self._get_summaries(grads_and_vars, add_grads=False)
> model_vars = BaseModelVars(global_step, train_op, train_summary_op, self.checkpoint_prefix)
> hooks = None
> if self.training_mode == TrainingMode.TF_ASYNC:
> ​    master = self.server.target
> elif self.training_mode == TrainingMode.TF_SYNC:
> ​    master = self.server.target
> ​    hooks = [self.optimizer.make_session_run_hook(is_chief=self.is_chief)]
> elif self.training_mode == TrainingMode.HOROVOD_SYNC:
> ​    master = ''
> ​    hooks = [hvd.BroadcastGlobalVariablesHook(0)]
> ​    session_conf.gpu_options.visible_device_list = str(hvd.local_rank())
> else:
> ​    raise Exception("training mode is not allowed: %d", self.training_mode)
> with tf.train.MonitoredTrainingSession(
> ​        master=master,
> ​        config=session_conf,
> ​        is_chief=self.is_chief,
> ​        checkpoint_dir=self.FLAGS.pretrained_model_dir,
> ​        hooks=hooks,
> ​        chief_only_hooks=[checkpoint_hook]) as sess:
> ​    self._train(sess, model_vars)

Since I am quite a newbie, I can't figure out how to fix this. So I'm looking forward to your help.

Many thanks.

**Comment by joanlynnlove** [comment]
Thanks, and yes, I have downloaded mnist dataset manually and passed that example.
My container can't connect to internet due to our company's security settings. Therefore I guess it may lack some components that horovod needs.

**Comment by joanlynnlove** [comment]
Thanks, and yes, I have downloaded mnist dataset manually and passed that example.
My container can't connect to internet due to our company's security settings. Therefore I guess it may lack some components that horovod needs.

**Comment by joanlynnlove** [comment]
Thanks, and yes, I have downloaded mnist dataset manually and passed that example.
My container can't connect to internet due to our company's security settings. Therefore I guess it may lack some components that horovod needs.

**Comment by joanlynnlove** [comment]
Thanks, and yes, I have downloaded mnist dataset manually and passed that example.
My container can't connect to internet due to our company's security settings. Therefore I guess it may lack some components that horovod needs.

**Comment by tgaddair** [comment]
I see. To clarify, you're saying that you were successfully able to run the example training script with _two_ processes without any hanging?  Specifically:

```
num_workers=2; mpirun --allow-run-as-root -np $num_workers -H localhost:$num_workers -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib python **tensorflow_mnist.py**
```

If that is true, then the issue would most likely not be with Horovod, but somewhere in the script.

One thing I noticed is that you explicitly set the "chief" in the MonitoredTrainingSession:

```
with tf.train.MonitoredTrainingSession(..., is_chief=self.is_chief, chief_only_hooks=[checkpoint_hook], ...)
```

This could be the source of the error. In Horovod, all workers should be a "chief", meaning that they initialize variables and do other init stuff. I would try removing those parameters, and instead set the checkpoint directory to `None` for workers other than rank 0, as done in the [tensorflow_mnist.py](https://github.com/uber/horovod/blob/master/examples/tensorflow_mnist.py) example.

Please give that a shot and let me know how it works.

**Comment by joanlynnlove** [comment]
@tgaddair Yeah! That's it! I tried that as you have advised and it works now! I have chief_only_hooks, so I rely on the parameter is_chief.
I was not running tensorflow_mnist.py with mpirun. However it also worked when I was using mpirun. 
Now it's time for me to have some further work with horovod.
Thanks! @tgaddair, I really appreciate your help!
